src/intra_edge.rs

```diff
@@ -38,182 +70,334 @@ pub struct EdgeBranch {
                 }) | (if !(n == 0 || (n == 2 && left_has_bottom)) {
                     0 as EdgeFlags
                 } else {
-                    EDGE_I444_LEFT_HAS_BOTTOM
-                        | EDGE_I422_LEFT_HAS_BOTTOM
-                        | EDGE_I420_LEFT_HAS_BOTTOM
-                })) as EdgeFlags,
-            );
+                    EDGE_LEFT_HAS_BOTTOM
+                });
+                self.tip[tip.index as usize] = EdgeTip::new(edge_flags);
+                n += 1;
+            }
+        } else {
+            let mut n = 0;
+            while n < B as u8 {
+                let (child_branch, next) = indices.branch[bl as usize].pop_front();
+                indices.branch[bl as usize] = next;
+                branch.split[n as usize] = child_branch;
+                (self, indices) = self.init_mode_node(
+                    child_branch,
+                    bl + 1,
+                    indices,
+                    !(n == 3 || (n == 1 && !top_has_right)),
+                    n == 0 || (n == 2 && left_has_bottom),
+                );
+                n += 1;
+            }
+        };
+        self.branch[branch_index.index as usize] = branch;
+        (self, indices)
+    }
+
+    const fn init(mut self) -> Self {
+        let mut indices = EdgeIndices {
+            branch: [EdgeIndex {
+                index: 0,
+                kind: EdgeKind::Branch,
+            }; 3],
+            tip: EdgeIndex {
+                index: 0,
+                kind: EdgeKind::Tip,
+            },
+        };
+
+        let sb128 = SB128 as u8;
+
+        let mut bl = BL_128X128;
+        while bl <= BL_32X32 {
+            indices.branch[bl as usize].index = level_index(bl + sb128);
+            bl += 1;
+        }
+
+        let bl = if SB128 { BL_128X128 } else { BL_64X64 };
+        (self, indices) = self.init_mode_node(EdgeIndex::root(), bl, indices, true, false);
+
+        let mut bl = BL_128X128;
+        while bl <= BL_32X32 {
+            let index = indices.branch[bl as usize].index;
+            if index != 0 {
+                assert!(index == level_index(1 + bl + sb128));
+            }
+            bl += 1;
         }
-    } else {
-        let nwc_children = slice::from_raw_parts_mut(mem.nwc[bl as usize], nwc.split.len());
-        mem.nwc[bl as usize] = mem.nwc[bl as usize].offset(nwc_children.len() as isize);
-        for (n, (split, nwc_child)) in iter::zip(&mut nwc.split, nwc_children).enumerate() {
-            *split = &mut nwc_child.node;
-            init_mode_node(
-                nwc_child,
-                bl + 1,
-                mem,
-                !(n == 3 || (n == 1 && !top_has_right)),
-                n == 0 || (n == 2 && left_has_bottom),
-            );
+        assert!(indices.tip.index == self.tip.len() as u8);
+
+        self
+    }
+
+    /// Check that all indices are in bound so that bounds checks are not needed at runtime.
+    const fn check_indices(self) -> Self {
```

Comment:
huh, the optimizer takes that into account? that is extremely neat!

Reply:
Nope lol.  I did it manually with an `unsafe`.  The optimizer is generally pretty bad at this kind of non local optimization and understanding any invariants on fields.